2016/proposals/narayan-aditya-visualizer.md

```diff
@@ -0,0 +1,385 @@
+
+##### PROS
+1. Compatible across all database engines supported by Django.
+2. We can query across the the Foreign Keys between the tables.
+
+##### CONS
+> In nearly every case, abstract inheritance is a better approach for the long term. I’ve seen more than few sites crushed under the load introduced by concrete inheritance, so I’d strongly suggest that Django users approach any use of concrete inheritance with a large dose of skepticism.
+
+-Jacob Kaplan-Moss
+
+The left joins that exist between the tables under multi-table
+inheritance cause a lot of database hit when Django fetches a
+row.
+
+#### Proposal 2
+
+This schema uses abstract inheritance with the use of
+GenericForeignKeys over multi-table inheritance.
+The base _Question_ and _Answer_ tables do not exist and the
+_Response_ table has two GenericForeignKeys, one that points to
+any of the Question table, while the other GenericForeignKey
+points to any of the corresponding _Answer_ tables.
+
+![Proposal-2](https://cloud.githubusercontent.com/assets/6830800/13756243/fb11080c-ea44-11e5-8790-7208e61c46f0.png)
+
+##### PROS
+
+None
+
+##### CONS
+
+1. We cannot run queries across the GenericForeignKey.
+   We cannot filter responses by the particular answer to a question.
+2. To allow the app to filter responses by a particular field,
+   we will have to explicitly save the field
+   (like the answer to a particular question)
+   in another column of the *Response* table.
+   This introduces a redundancy.
+
+#### Proposal 3
+
+We stores the fields within the Answer table as a serialized dictionary
+in a TextField of the column json.
+
+![Proposal-3](https://cloud.githubusercontent.com/assets/6830800/13775856/ddf50358-eacc-11e5-8320-8ae205014310.png)
+
+##### PROS
+
+1. Simple schema.
+
+##### CONS
+
+1. No SQL validation across the fields of the *Answer* table stored
+   within the TextField. The fields are saved as strings.
+2. We incur an overhead of parsing and serializing the fields
+   on every database hit.
+3. To allow the app to filter responses by a particular field,
+   we will have to explicitly save the field
+   (like the answer to a particular question)
+   in another column of the *Response* table.
+   This introduces a redundancy.
+
+#### Proposal 4
+
+We use the HStoreField that PostgreSQL provides to store the fields
+of the *Answer* tables.
+
+![Proposal-4](https://cloud.githubusercontent.com/assets/6830800/13776834/47d28e4e-ead2-11e5-8f6d-bebeb972ef1a.png)
+
+##### PROS
+
+1. Simplest database schema to understand and work with.
+2. Allows us to query responses by the answer.
+   `Response.objects.filter(answer__data__text='Workshop A')`
+
+##### CONS
+
+1. This schema can only be realized on PostgreSQL database,
+   on both development and production environments.
+   No equivalent SQLite solution exists.
+2. No SQL validation across the fields of the *Answer* table stored
+   within the HStoreField. The fields are saved as strings.
+
+#### Summary
```

Comment:
If you got in, we'd spend some time, with Greg and/or Trevor, to discuss these approaches and select the best one or the first one to start with.

In a nutshell – if I'm wrong please correct me – proposals 1-2 have the most complicated table structure, and proposals 3-4 store answers in single field; the difference between the latter is that in proposal 3 it's a normal `text` field and it would store JSON, while in proposal 4 it's a special PostgreSQL field `HStore` we could use.